File: src/c/ch01/1.1/l0/exercises_starter.c
#include <stdio.h>
#include <stdlib.h>
#include "dec_to_bin.h"
#include "bin_to_dec.h"

static void ex1(void) {
    char* s = dec_to_bin(19);
    printf("Exercise 1 (19 -> binary): %s\n", s);
    free(s);
}

static void ex2(void) {
    unsigned int out;
    if (bin_to_dec("10101", &out)) {
        printf("Exercise 2 ('10101' -> dec): %u\n", out);
    }
}

static void ex3(void) {
    char* s = dec_to_bin(27);
    printf("Exercise 3 (27 -> binary): %s\n", s);
    free(s);
}

static void ex4(void) {
    unsigned int out;
    if (bin_to_dec("111111", &out)) {
        printf("Exercise 4: 111111 -> %u == 63? %s\n", out,
               out == 63 ? "true" : "false");
    }
}

static void ex5(void) {
    puts("Exercise 5: Binary fits hardware with two stable states (0/1).");
}

int main(void) {
    ex1();
    ex2();
    ex3();
    ex4();
    ex5();
    return 0;
}

File: src/go/ch01/1.1/l0/exercises_starter.go
package main

import "fmt"

func runExercises() {
	// Exercise 1: 19 -> binary
	fmt.Println("Exercise 1 (19 -> binary):", DecToBin(19)) // expect 10011

	// Exercise 2: "10101" -> decimal
	if v, err := BinToDec("10101"); err == nil {
		fmt.Println("Exercise 2 ('10101' -> dec):", v) // expect 21
	} else {
		fmt.Println("Exercise 2 error:", err)
	}

	// Exercise 3: 27 -> binary
	fmt.Println("Exercise 3 (27 -> binary):", DecToBin(27)) // expect 11011

	// Exercise 4: verify 0b111111 == 63
	if v, err := BinToDec("111111"); err == nil {
		fmt.Printf("Exercise 4: 111111 -> %d == 63? %v\n", v, v == 63)
	} else {
		fmt.Println("Exercise 4 error:", err)
	}

	// Exercise 5: why binary? (short explanation)
	fmt.Println("Exercise 5: Electronics are reliable with two states (0/1);",
		"binary is simpler and more robust for circuits than base 10.")
}
